cpp/src/arrow/adapters/avro/adapter.h

```diff
@@ -0,0 +1,93 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_AVRO_CONVERTER_H
+#define ARROW_AVRO_CONVERTER_H
+
+#include <cstdint>
+#include <memory>
+#include <string>
+
+#include "arrow/builder.h"
+#include "arrow/buffer.h"
+#include "arrow/memory_pool.h"
+#include "arrow/status.h"
+#include "arrow/table.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit-util.h"
+#include "arrow/util/macros.h"
+#include "arrow/util/visibility.h"
+
+#include "avro.h"
+
+namespace arrow {
+
+namespace adapters {
+
+namespace avro {
+
+/// Base class for reading from an avro datasource.
+//
+/// This class provides a facilities for incrementally building the null bitmap
+/// (see Append methods) and as a side effect the current number of slots and
+/// the null count.
+class ARROW_EXPORT AvroArrowReader {
+ public:
+  explicit AvroArrowReader(MemoryPool* pool)
+      : pool_(pool) {
+  }
+
+  virtual ~AvroArrowReader() = default;
+
+  Status ReadFromAvroFile(avro_file_reader_t* file_reader,
+                          std::shared_ptr<Array>* out);
+ protected:
+  AvroArrowReader() {
```

Comment:
I would suggest following a private implementation pattern, so here we would have

```
class AvroStreamReaderImpl;
std::unique_ptr<AvroStreamReaderImpl> impl_;
```

You can find lots of examples of this throughout the codebase to get you started

Reply:
Would a forward declaration of `avro_file_reader_t` be unacceptable here? It doesn't look there's any state in the pimpl class and it would be less complex to not use it if it isn't necessary.

Reply:
This class needs to initialize `avro_file_reader_t` from one of Arrow's IO interfaces (like `io::InputStream` in the PIMPL)